src/storage/src/source/generator/clock.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_ore::now::NowFn;
+use mz_repr::{Datum, Row};
+use mz_storage_types::sources::load_generator::{Event, Generator};
+use mz_storage_types::sources::MzOffset;
+use std::iter;
+use std::time::Duration;
+
+pub struct Clock {
+    pub interval: Duration,
+}
+
+impl Generator for Clock {
+    fn by_seed(
+        &self,
+        now: NowFn,
+        _seed: Option<u64>,
+        resume_offset: MzOffset,
+    ) -> Box<dyn Iterator<Item = (usize, Event<Option<MzOffset>, (Row, i64)>)>> {
+        let tick_ms: u64 = self.interval.as_millis().try_into().unwrap();
+        let mut last_offset = resume_offset.checked_sub(MzOffset::from(1));
+        Box::new(
+            iter::from_fn(move || {
+                let current_offset = MzOffset::from(now() / tick_ms);
```

Comment:
I think this line here will make the output of the generator non-definite, in the sense that two instances of this generator resuming at the same `resume_offset` will produce different updates depending on what `now()` returns. I think what you want to do is:
1. Define a *pure* function with an `(offset) -> timestamp` signature that computes something like `start_timestamp + offset * interval`, where `start_timestamp` is a durably recorded timestamp of the source. You can record that during purification with a generator option.
2. Given that the generator framework yields the generator every time it produces a progress message you can use `now()` to compute a `next_progress: Option<MzOffset>` value that instructs the `from_iter` implementation what range of offsets it must produce as events before producing a progress message.
3. Crucially, the generator emits the corresponding `Event::Message` for every offset starting at `resume_offset`. i.e no offset should be skipped from the output.

This is what I have in mind in pseudocode:

```rust
let offset_to_timestamp = |offset| self.start_timestamp + offset * self.tick_ms;
let mut upper = resume_offset;
from_fn(|| {
    let prev_upper = std::mem::replace(
        &mut upper,
        std::cmp::max(upper, MzOffset::from(now() / tick_ms) + 1)
    );
    (prev_upper..upper)
        .flat_map(|off| off
            .checked_sub(1)
            .map(|off| Event::Message(offset_to_timestamp(off), -1));
            .into_iter()
            .chain([Event::Message(offset_to_timestamp(off), 1)])
        )
        .chain([Event::Progress(upper)])
}).flatten()
```

Reply:
Mmm... I can see why you'd want a stream that accumulates correctly at every frontier, not just the emitted progress frontiers.

I'll follow up on this in a couple days when I have time to refresh my memory on purification...

Reply:
> I think this line here will make the output of the generator non-definite, in the sense that two instances of this generator resuming at the same `resume_offset` will produce different updates depending on what `now()` returns.

That's fine though, right? We reclock generators, so it's okay that the generator itself is nondefinite.

I think whether or not we guarantee that we produce an update for every tick or not is a product call that could go either way—possibly something that we expose to the user via an option. Personally the use cases I have in mind for this would prefer to skip times if the load generator goes down, rather than replaying all the intermediate times.

Reply:
If we do want to go the "produce all intermediate messages" route:

> 1. Define a _pure_ function with an `(offset) -> timestamp` signature that computes something like `start_timestamp + offset * interval`, where `start_timestamp` is a durably recorded timestamp of the source. You can record that during purification with a generator option.

I think we could just use the existing `AS OF` load generator option. If `AS OF` isn't explicitly specified, during purification, you'd take the current time, convert it to a tick offset, and install it as the `AS OF`.

Reply:
Anyway, I think we should shake out the product question of whether to produce all intermediate messages or not before you sink time into adding the implementation, @bkirwi! If the current implementation is not incorrect, I think there's a strong case to be made for skipping times being the default behavior.

Reply:
> That's fine though, right? We reclock generators, so it's okay that the generator itself is nondefinite.

No, sources must be definite even in their source timestamp domain (in this case mzoffset). reclocking can't turn a non-definite collection into a definite one

Reply:
> I think we should shake out the product question of whether to produce all intermediate messages or not before you sink time into adding the implementation

The intermediate timestamps must be produced here (for definiteness reasons) but if a bunch of them get reclocked at a single timestamp then the user will see their consolidated value, i.e the clock jumping forward by more than one tick just like the current impl. It's not that the user will observe all these ticks.

Reply:
> No, sources must be definite even in their source timestamp domain (in this case mzoffset). reclocking can't turn a non-definite collection into a definite one

But, our existing sources don't always work like this! For example, reading the same Kafka topic from two different processes can produce different results due to retention policies or compaction occurring. What we're doing here is pretty much exactly what would happen if you had a Kafka source with compaction enabled where you published the current time periodically, always with the same key (e.g., `clock`).

Reply:
> but if a bunch of them get reclocked at a single timestamp then the user will see their consolidated value, i.e the clock jumping forward by more than one tick just like the current impl. It's not that the user will observe all these ticks.

I was thinking of the user here as downstream matviews or `SUBSCRIBE` statements, not users manually issuing `SELECT` statements. In particular, emitting every tick will force downstream matviews to recompute all the intermediate times, rather than jumping straight to the current time.

Reply:
I hate to say it but I think Petros is right! 😅

I think the practical issue is that storage's Persist sink assumes that the collection accumulates correctly at every timestamp after the collection's initial since, which is why it's fine to discard any updates before the observed upper. However, my load-gen output would only accumulate correctly at the explicit progress messages. (This is relatively benign for an append-only Kafka source, since the worst that can happen is a missing append. But for this source I believe it could result in either a negative accumulation or some timestamp never being retracted.)

Reclocking mitigates this, because messages at distinct wall-clock timestamps can get reclocked to the same mztime, so things consolidate out / see fewer frontier advancements when the source is catching up. Tested locally with the updated code; taking the source down for three minutes results in only one post-reclocking update:

```
materialize=> copy(subscribe to (select time from minute)) to stdout;
1719590980192	1	2024-06-28 16:09:00+00
1719591138000	-1	2024-06-28 16:09:00+00
1719591138000	1	2024-06-28 16:12:00+00
```

Reply:
> I think the practical issue is that storage's Persist sink assumes that the collection accumulates correctly at every timestamp after the collection's initial since, which is why it's fine to discard any updates before the observed upper. However, my load-gen output would only accumulate correctly at the explicit progress messages. (This is relatively benign for an append-only Kafka source, since the worst that can happen is a missing append. But for this source I believe it could result in either a negative accumulation or some timestamp never being retracted.)

I'm still not quite tracking what goes wrong, but I think that's okay because...

> Reclocking mitigates this, because messages at distinct wall-clock timestamps can get reclocked to the same mztime, so things consolidate out / see fewer frontier advancements when the source is catching up. Tested locally with the updated code; taking the source down for three minutes results in only one post-reclocking update:

..that mitigates my concern! I forgot this was all happening upstream of reclocking, so until we do https://github.com/MaterializeInc/materialize/issues/21428 the big pile of updates will all get consolidated together before they're made visible to the users.

Reply:
> But, our existing sources don't always work like this! For example, reading the same Kafka topic from two different processes can produce different results due to retention policies or compaction occurring. 

If this ever happens the contents of the source are undefined. In fact on start up we should assert that our resume upper is beyond the low watermark of the topic and error if that's not the case.

> I'm still not quite tracking what goes wrong, but I think that's okay because...

This is the wrong question to ask because it might be the case that nothing goes wrong today because no one is relying on definiteness in just the right way to cause an observable issue. But we do build everything around sources with that property assumed true so if we want to ensure that future local reasoning is correct we must adhere to the contract. We may want to question the value of definiteness but it must be a decision that applies to the entire system, not just one source implementation.

All that said here is concrete example of what can go wrong today *only for illustration purposes*. 

Let's assume that we run the clock generator as originally written produces this output:

```
(12:00, 0, +1)
// progress [1]
(12:00, 50, -1)
(12:05, 50, +1)
// progress [51]
```

We record the following remap collection:

```
(1, t1, +1)
(1, t2, -1)
(51, t2, +1)
```

But because of reasons, we only manage to commit to persist the data until t1. So the persist shard contains:

```
(12:00, t1, +1)
```

Now we crash and restart. The new clock instance gets instantiated with a resume offset of 1 and it now produces these updates instead of the original ones (because it's not definite):

```
(12:00, 20, -1) // correctly subtract the `last_offset`, which we know from resume offset
(12:02, 20, +1)
// progress [21]
(12:02, 70, -1)
(12:07, 70, +1)
// progress [71]
```

Since the clock advanced to offset 71 and we have a binding for offset 51 we record these additional updates in the persist shard:

```
(12:00, t2, -1)
(12:02, t2, +1)
```

We crash again and restart. Now the resume offset is 51 (this is what t2 maps to) and the contents of the collection are:

```
(12:00, t1, +1)
(12:00, t2, -1)
(12:02, t2, +1)
```

It is now game over. The clock being resumed at offset 51 will produce a negative diff for `12:05`, which was never written.

Reply:
> This is the wrong question to ask because it might be the case that nothing goes wrong today because no one is relying on definiteness in just the right way to cause an observable issue... We may want to question the value of definiteness but it must be a decision that applies to the entire system, not just one source implementation.

Oh, sorry, what I meant is that even though I didn't see the hazard, I was (and still am) totally comfortable proceeding with your suggested implementation which guarantees definiteness. I was worried that your proposal would cause unnecessary work for downstream matviews after source downtime, but I see now that's not the case.

> All that said here is concrete example of what can go wrong today only for _illustration purposes_.

But, nonetheless, thank you for this! I get it now. And sorry, I was trying to spare you from having to write this up just for my sake, but I failed to communicate that clearly.

My quick summary of the hazard is this: the `resume_offset` is not necessarily the last offset committed to the data shard; it is merely the last offset committed to the reclocking shard.

> If this ever happens the contents of the source are undefined. In fact on start up we should assert that our resume upper is beyond the low watermark of the topic and error if that's not the case.

The fact that you consider this "undefined behavior" worries me mildly, fwiw! Kafka can 100% delete the record at an offset between when we reclock that offset and when when we write that offset to persist, if we crash at the wrong moment as in your illustrative example. I don't think checking the low watermark of the topic is sufficient. I think both Kafka retention policies and compaction policies can delete data from anywhere in a topic. 

I think @bkirwi is bang on here too that in practice this won't cause issues:

> This is relatively benign for an append-only Kafka source, since the worst that can happen is a missing append. But for this source I believe it could result in either a negative accumulation or some timestamp never being retracted.

(I think the same applies to Kafka upsert sources, too; the worst that can happen is that you miss an update, but that should only occur because there is another more recent update that you are about to see.)

Reply:
> Kafka can 100% delete the record at an offset between when we reclock that offset and when when we write that offset to persist, if we crash at the wrong moment as in your illustrative example.

The mental model for kafka is that when a source is created we durably record the current `[low, high]` offset range of the topic and we set up ingestion such that it only reads offsets starting at `low`, and that the initially observed `[low, high]` range is produced in the form of a snapshot at a single timestamp. Any dataflow attempting to ingest this topic can assert that the messages at offset low or above are present by checking the low watermark. 

This is relatively fine if we're talking about a new topic whose retention policy hasn't yet kicked in, but the race is indeed tighter if a source targeted against a topic that retention has already started to delete data. In that case as you said we can have records missing between the time we observe the snapshot's low watermark on creation and attempting to ingest it at some later time. We don't currently report an error in that case but we should, according to the model at least. We have some freedom of choice here though because if on source creation we know we're dealing with a topic that has had retention kicked in we can select an initial `low` watermark for the snapshot that is greater than the low watermark of the topic. The greater the distance that more time we have to win the race, at the expense of ignoring a prefix of data (that would get deleted anyway in some time). 

There is definitely an impedance mismatch here and it's unfortunate Kafka doesn't have the concept of read holds for retention. 

> I think @bkirwi is bang on here too that in practice this won't cause issues:

I think it is true that the append-only case does not cause issues today even there I can see problems in the future. For example in a replicated ingestion scenario where two dataflows are running concurrently if they observe a different set of updates upstream they might start oscillating in their persist sink where they constantly attempt to cancel each other's "extra" update out.

In the case of upsert if our read frontier ever falls behind the compaction frontier we start having correctness property #2 violations.

Reply:
> The mental model for kafka is that when a source is created we durably record the current [..]

Just to be clear, we don't do that currently. But it's against that model that I validate whether other things are correct

Reply:
> I think it is true that the append-only case does not cause issues today even there I can see problems in the future ... In the case of upsert if our read frontier ever falls behind the compaction frontier we start having correctness property #2 violations.

Yeah, I'm totally comfortable with allowing these anomalies to occur for Kafka sources in the case of racing with retention or compaction policies. There's just no way to completely eliminate these anomalies without proper read holds.

Where I'd get worried is if racing with retention or compaction policies caused true undefined behavior in the C sense of the term—i.e., if the data in the source became arbitrarily incorrect, rather than having, uhm, well-defined incorrectness.

> For example in a replicated ingestion scenario where two dataflows are running concurrently if they observe a different set of updates upstream they might start oscillating in their persist sink where they constantly attempt to cancel each other's "extra" update out.

I think we'll have a solve for this as part of the 0dt work, happily! We need to teach sources to, while in read only mode, listen to changes that the old version is making and apply those changes to its local RocksDB hashmap. I think the same technique will apply in the replicated case, too. The upsert source can keep track of which entries in the hashmap have been committed to the data shard and which haven't. On a compare-and-append failure, the upsert source should dump whatever entries in the hashmap weren't committed, and then replay the entries from the data shard to catch back up.

Reply:
> Where I'd get worried is if racing with retention or compaction policies caused true undefined behavior in the C sense of the term—i.e., if the data in the source became arbitrarily incorrect, rather than having, uhm, well-defined incorrectness.

This is a race of which the outcome is known to us (the low watermark), so one way to mitigate this worry is to be more strict, which is what I'm advocating for. I don't think we want to be in the business of reasoning whether a slightly non-definite source happens to work correctly because of how we implemented 0dt or whatever else we do in the future. e.g does that slight non-definiteness work correctly with a feature where we add an expiration based on the data? Could that maybe lead to negative retractions? Who knows. Sure, we can maybe figure out in that case too but this is similar to using a database in read committed mode which works fine until it doesn't.

> Not sure how intelligible that was—I'm short on time!—but we'll certainly discuss more when we're at that point of the 0dt work.

Made total sense!

Reply:
> This is a race of which the outcome is known to us (the low watermark), so one way to mitigate this worry is to be more strict, which is what I'm advocating for. I don't think we want to be in the business of reasoning whether a slightly non-definite source happens to work correctly because of how we implemented 0dt or whatever else we do in the future. e.g does that slight non-definiteness work correctly with a feature where we add an expiration based on the data? Could that maybe lead to negative retractions? Who knows. Sure, we can maybe figure out in that case too but this is similar to using a database in read committed mode which works fine until it doesn't.

I can guarantee you though that we have customers who absolutely do not want their source to get bricked because of one missing Kafka record. I agree with you in principle that it would be nice if the default were to throw an error if we ever discovered a missing Kafka offset, but I think in practice that behavior would be extremely irritating for a large chunk of our customers.

Reply:
I'm going to mark this resolved since I think we've settled the relevant bits for this PR, but happy to pick this thread up internally if y'all are interested!